### DISCUSSION TITLE
Testing whether an Hash is empty without extra functions from modules (like stdlib). Puppet 4.10

### DISCUSSION THREAD

**Discussion by pier4r** [puppetlabs/community#5] [Q&A] [answered]
Disclaimer: if this post/place is going to be removed let me/us users know first so we have the chance to post the content elsewhere for other puppet users.

At the moment on the systems there is 4.10 available. Now in some cases we have the possibility to use functions from modules, like standardlib, but if possible I would like to avoid to rely too much on extra things that may change/may be unavailable in some circumstances (partial installations, puppet apply, modules that may avoid using other modules, etc..)

Thus I have a fact that is an **Hash**, say `$facts[test_fact]` and I'd like to test whether it is empty.

With `stdlib` it would be `$facts[test_fact].size > 0` .

What if I do not use it? The best replacement I could identify, [using the manual](https://puppet.com/docs/puppet/4.10/lang_data_hash.html), is:

```
$facts[test_fact] =~ Hash[Scalar, Data, 1]
```

It seems to work and I do not find it that bad considering that it saves the dependency on a module.

Is there anything better?

**Comment by pier4r** [comment]
from https://puppetcommunity.slack.com/archives/C0W298S9G/p1611326656098700

## Yury B.
How about if `$fact[‘xxx’] == {}`?

## Pierfrancesco A.
true (duh that was obvious, thank you!) , that would do it too! I want a "not empty" thus maybe something like:  
`if (! $fact['xxx'] == {} )`  
Would it work?  
On the other hand if I want a certain size (say, at least 2) still I would use  
`Hash[Scalar, Data, 2]`  
I am still not sure where there is another way. 

## Yury B.
`if $fact[‘xxx’] != {}` I guess
Though your way should work too
In modern puppet I'd prefer using empty() function though

**Comment by pier4r** [comment, marked as answer]
### Pierfrancesco 
I was thinking, could it be that
`if $fact['xxx'] != {}`
is true also if the fact is undef ? (same for  `if (! $fact['xxx'] == {}` )
and thus  `Hash[Scalar, Data, 1]` is safer?
New

### helindbe
the first is false if, and only if the value is a hash and is empty, and true for all other values. (edited) 
the second is the same

What you want to use depends on what you really want to test for. The `empty()` function is true for empty array, hash, string, and for undef. If you are interested more strictly in an empty array or hash, you can use `=~ Collection[0,0]` (empty) or `=~ Collection[1]` (at least one - i.e. not empty).
You can also use `Hash[0,0]` for empty hash.

If you want to also ignore hashes that does not have Scalar key and Data values then do use `Hash[Scalar,Data, 1]` (i.e. "hash has at least one such element"). (edited) 

### Pier
Didn't know that I could simply use Hash[1] or Collection[1] (reading the 4.10 manual it seemed that I needed to give two values first then the min size)

And I also didn't know that empty (that should be also in stdlib) is true also for undef. That is a tad less safer in my case.

**Reply by hlindberg** [reply]
You cannot use `Hash[0]` (single value), but you can omit the type if you give two values. So `Hash[0,0]` is the type for empty hash. For `Collection` there is only one parameter - the size, so `Collection[0]` is allowed.

**Reply by hlindberg** [reply]
If you only give one value (min) (if allowed to give only one) the max value is `default` which means "open ended".
